### PR TITLE
osd: avoid unexpected filestore osd when no new devices are configured

### LINKED ISSUE #2445: [Ceph] Unspecified filestore osds are created on operator recreation
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

Not specified filestore osds are created

cephclusters.ceph.rook.io rook-ceph:
```
spec:
  cephVersion:
    image: ceph/ceph:v13.2.2-20181023
  dashboard:
    enabled: true
  dataDirHostPath: /var/lib/rook
  mon:
    allowMultiplePerNode: false
    count: 3
  network:
    hostNetwork: false
  rbdMirroring:
    workers: 0
  storage:
    config:
      databaseSizeMB: "1024"
      journalSizeMB: "1024"
      storeType: bluestore
    nodes:
    - config:
        storeType: bluestore
      devices:
      - FullPath: ""
        config: null
        name: nvme1n1
      name: server01
      resources: {}
    - config:
        storeType: bluestore
      devices:
      - FullPath: ""
        config: null
        name: nvme1n1
      name: server02
      resources: {}
    - config:
        storeType: bluestore
      devices:
      - FullPath: ""
        config: null
        name: sda
      - FullPath: ""
        config: null
        name: sdb
      name: server03
      resources: {}
    useAllDevices: false
```

Instead of 4 osds 6 are created:

```
ID CLASS WEIGHT  TYPE NAME         STATUS REWEIGHT PRI-AFF 
-1       8.76755 root default                              
-3       7.27399     host server03                      
 0   hdd 3.63699         osd.0         up  1.00000 1.00000 
 2   hdd 3.63699         osd.2         up  1.00000 1.00000 
-2       0.90659     host server02                         
 4       0.44260         osd.4         up  1.00000 1.00000 
 1  nvme 0.46399         osd.1         up  1.00000 1.00000 
-4       0.58698     host server01                             
 5       0.12299         osd.5         up  1.00000 1.00000 
 3  nvme 0.46399         osd.3         up  1.00000 1.00000
```
OSDs 4 and 5 are unexpected

**Expected behavior:**

Respect of the spec.storage.useAllDevices directive

**How to reproduce it (minimal and precise):**

Specify bluestore OSDs and set spec.storage.useAllDevices = false

**Environment**:
* OS (e.g. from /etc/os-release):
```
NAME="Ubuntu"
VERSION="18.04.1 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.1 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
* Kernel (e.g. `uname -a`):
`Linux server01 4.15.0-42-generic #45-Ubuntu SMP Thu Nov 15 19:32:57 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
* Cloud provider or hardware configuration: 
Bare Metal
* Rook version (use `rook version` inside of a Rook Pod):
```rook: v0.9.1```
* Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T21:04:45Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T20:56:12Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
```

* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)): 
HEALTH_OK

### PR DESCRIPTION
**Pull request by clkao**
Signed-off-by: Chia-liang Kao <clkao@clkao.org>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #2445

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

### CODE REVIEW

**Comment by clkao** [comment]
This appears to be a regression introduced in 0.9.0 by https://github.com/rook/rook/commit/1389cc56db4eac538823e5bf688486c9460b61c3

Might worth backporting to release-0.9.

**Comment by travisn** [comment]
@clkao The integration tests are failing. It looks like this change is always preventing the default test OSD from being created. I'd like to understand the repro better, see my comments on the original issue. Thanks for looking into this, I overlooked it after the break.

**Comment by clkao** [comment]
@travisn this is weird, the integration tests pass for me locally.

jenkins shows `Error from server (AlreadyExists): error when creating "STDIN": object is being deleted: customresourcedefinitions.apiextensions.k8s.io "cephclusters.ceph.rook.io" already exists` so I am guessing it's running tests consecutively and didn't wait for cleanups.

**Comment by immesys** [comment]
Will this change still prevent a default directory OSD being created if no devices are specified? What happened to me is that my config specified useAllDevices = true and useAllNodes = true but some servers had no extra disks attached yet, so they created directory OSDs. 

I ask because I am not sure if `len(agent.devices)` will be `>0` in that case. I don't know the codebase so I could well be wrong.

**Comment by clkao** [comment]
according to https://github.com/rook/rook/blob/master/cmd/rook/ceph/osd.go#L254, deviceFilter is implemented as an agent entry with `isFilter` flag. so I'd assume useAllDevices being a `*` filter. That is, in this case the deviceSpecified should be 1 and rook should not create the default filestore osd.

**Comment by clkao** [comment]
It turns out when empty --data-devices passed to osd prepare, we get a ghost device `[{Name: OSDsPerDevice:1 IsFilter:false}]` due to golang's strings.Split semantics. That makes the the prepare job always think some devices are specified and it should never create a default filestore osd.

File: cmd/rook/ceph/osd.go [hunk 1]
```diff
@@ -310,6 +310,11 @@ func commonOSDInit(cmd *cobra.Command) {
 //   sda,sdb,nvme01:5
 func parseDevices(devices string) ([]osddaemon.DesiredDevice, error) {
 	var result []osddaemon.DesiredDevice
+
+	if devices == "" {
```
[the remark this answers is not in the document]
Comment (clkao) [reply cmd/rook/ceph/osd.go]: The ghost device issue only manifests after the logic for [checking deviceSpecified is changed](https://github.com/rook/rook/pull/2546/files#diff-d1ca60bbccae22948f1847a6b9747cd5R141), as fallback default osd is only created when no devices are specified.  Now when the user launch cephcluster with the default setting, the ghost device indicates deviceSpecified, as if the user specified some device and did not want to default osd to be created, which is probably not our intended behaviour.

File: tests/integration/zblock_mount_unmount_test.go [hunk 2]
```diff
@@ -85,7 +85,7 @@ func (s *BlockMountUnMountSuite) SetupSuite() {
 	s.pvcNameRWO = "block-persistent-rwo"
 	s.pvcNameRWX = "block-persistent-rwx"
 	useHelm := false
-	useDevices := true
+	useDevices := false
```
[the remark this answers is not in the document]
Comment (clkao) [reply tests/integration/zblock_mount_unmount_test.go]: when we specify useDevices, this now means we don't want the default filestore osd and that's why the test failed in the initial PR.  i think for integration test we should probably test device and dir filestore separately and explicitly, as well as the logic determining the osd to provision with a combination of user options.

**Comment by travisn** [comment]
@clkao Since #2585 was merged, can you confirm if this PR is no longer needed?

**Comment by galexrt** [comment]
@clkao Ping. Can the PR be closed as written by @travisn?

**Comment by clkao** [comment]
@galexrt I believe so, as there's no longer the fallback osd

**Pull request closed without merging**